File: app/src/main/java/io/github/thang86/weathers/base/BaseDrawerActivity.java
package io.github.thang86.weathers.base;

import android.content.Intent;
import android.view.View;
import android.widget.ImageView;

import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;

import com.google.android.material.navigation.NavigationView;

import io.github.thang86.weathers.R;
import io.github.thang86.weathers.activity.search.SearchAreaActivity;

/**
 * ThangTX2
 */
public abstract class BaseDrawerActivity extends BaseActivity {
    private DrawerLayout mDrawer;
    private NavigationView mNavigationView;
    private ImageView mBtnSeach;
    private static final int SEARCH_AREA_REQUEST_CODE = 1;

    @Override
    public void setContentView(int layoutResID) {
        super.setContentView(layoutResID);
        bindViews();
        setNavigationItemSelected();
    }

    @Override
    protected void bindViews() {
        super.bindViews();
        mDrawer = findViewById(R.id.drawer_layout);
        mBtnSeach = findViewById(R.id.btn_search_location);
        mDrawer.setDrawerLockMode(DrawerLayout.LOCK_MODE_LOCKED_CLOSED);
        mNavigationView = findViewById(R.id.nav_view);
        mBtnSeach.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity();
            }
        });
    }

    private void startActivity() {
        Intent searchLocationIntent = SearchAreaActivity.newIntent(getApplicationContext());
        startActivityForResult(searchLocationIntent, SEARCH_AREA_REQUEST_CODE);
    }

    @Override
    protected void setupToolbar() {
        super.setupToolbar();
        if (getToolbar() != null) {
            getToolbar().findViewById(R.id.image_other).setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    openCloseDrawer(false);
                }
            });
        }
    }

    /**
     * Open or Closed Navigation
     *
     * @param isBackPress
     */
    void openCloseDrawer(boolean isBackPress) {
        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        if (drawer.isDrawerOpen(GravityCompat.END)) {
            drawer.closeDrawer(GravityCompat.END);
        } else {
            if (!isBackPress) {
                drawer.openDrawer(GravityCompat.END);
            } else {
                super.onBackPressed();
            }
        }
    }

    @Override
    public void onBackPressed() {
        openCloseDrawer(true);
    }

    /**
     * Open or Closed Navigation
     */
    private void openCloseDrawer() {
        DrawerLayout drawer = findViewById(R.id.drawer_layout);
//        if (drawer.isDrawerOpen(GravityCompat.END)) {
//            drawer.closeDrawer(GravityCompat.END);
//        } else {
//            super.onBackPressed();
//        }
        if (drawer.isDrawerOpen(GravityCompat.END)) {
            drawer.closeDrawer(GravityCompat.END);
        } else {
            drawer.openDrawer(GravityCompat.END);
        }
    }

    @Override
    public boolean isNetworkConnected() {
        return super.isNetworkConnected();
    }

    public abstract void setNavigationItemSelected();
}

File: app/src/main/java/io/github/thang86/weathers/core/BaseDrawerActivity.kt
package io.github.thang86.weathers.core

import android.content.Intent
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import io.github.thang86.weathers.ui.activity.search.SearchAreaActivity
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.navigation_toolbar.*
import kotlinx.android.synthetic.main.toolbar.*

abstract class BaseDrawerActivity : BaseActivity() {

    companion object {
        private const val SEARCH_AREA_REQUEST_CODE = 1
    }

    override fun setContentView(layoutResID: Int) {
        super.setContentView(layoutResID)
        bindViews()
        setNavigationItemSelected()
    }

    fun bindViews() {
        drawerLayout.setDrawerLockMode(DrawerLayout.LOCK_MODE_LOCKED_CLOSED)
        btnSearchLocation.setOnClickListener {
            openCloseDrawer(false)
            startActivity()
        }
        btnPrever.setOnClickListener { drawerLayout.closeDrawers() }
    }

    override fun onResume() {
        super.onResume()
        drawerLayout.closeDrawers()
    }

    /**
     * Open SearchArea screen
     */
    private fun startActivity() {
        val searchLocationIntent: Intent = SearchAreaActivity.newIntent(applicationContext)
        startActivityForResult(searchLocationIntent, SEARCH_AREA_REQUEST_CODE)
    }

    override fun setupToolbar() {
        super.setupToolbar()
        if (toolbar != null) {
            image_other.setOnClickListener {
                openCloseDrawer(false)
            }
        }
    }

    /**
     * open or closed drawer layout
     *  if isBackPress == true when tab button back
     */
    private fun openCloseDrawer(isBackPress: Boolean) {
        if (drawerLayout.isDrawerOpen(GravityCompat.END)) {
            drawerLayout.closeDrawer(GravityCompat.END)
        } else {
            if (!isBackPress) {
                drawerLayout.openDrawer(GravityCompat.END)
            } else {
                super.onBackPressed()
            }
        }
    }

    override fun onBackPressed() {
        super.onBackPressed()
        openCloseDrawer(true)
    }

    abstract fun setNavigationItemSelected()

}